# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/base/tests/browser_lazyload_page_load_telemetry_iframe.js
# Commit: 1cc86398ced0
# Full Hash: 1cc86398ced034fafc66d94799f91221420e1229
# Author: Sean Feng <sefeng@mozilla.com>
# Date: 2021-02-18 04:16:06
# Regressor Bug: 1692350
# File Overlap Count: 1
# Description:
#   Bug 1692350 - Implement a telemetry probe to collect page load data for documents that use lazyload r=emilio,chutten
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D104912
# ==============================================================================

diff -r 1ca4c3365fae -r 1cc86398ced0 layout/base/tests/browser_lazyload_page_load_telemetry_iframe.js
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/layout/base/tests/browser_lazyload_page_load_telemetry_iframe.js	Wed Feb 17 22:25:02 2021 +0000
@@ -0,0 +1,48 @@
+const baseURL = getRootDirectory(gTestPath).replace(
+  "chrome://mochitests/content",
+  "http://example.com"
+);
+
+const testFileURL = `${baseURL}file_lazyload_telemetry.html`;
+
+const { TelemetryTestUtils } = ChromeUtils.import(
+  "resource://testing-common/TelemetryTestUtils.jsm"
+);
+
+function OtherLazyLoadDataIsReported() {
+  const snapshot = Services.telemetry.getSnapshotForHistograms("main", false)
+    .content;
+  return snapshot.LAZYLOAD_IMAGE_TOTAL;
+}
+
+function pageLoadIsReported() {
+  const snapshot = Services.telemetry.getSnapshotForHistograms("main", false)
+    .parent;
+  return snapshot.FX_LAZYLOAD_IMAGE_PAGE_LOAD_MS;
+}
+
+add_task(async function testTelemetryCollection() {
+  Services.telemetry.getHistogramById("FX_LAZYLOAD_IMAGE_PAGE_LOAD_MS").clear();
+
+  const testTab = await BrowserTestUtils.openNewForegroundTab(
+    gBrowser,
+    "data:text/html,<body><iframe src='" + testFileURL + "'></iframe></body>",
+    true
+  );
+
+  await BrowserTestUtils.waitForCondition(pageLoadIsReported);
+
+  gBrowser.removeTab(testTab);
+
+  // Running this test also causes some other LAZYLOAD related data
+  // to be collected. Wait for them to be collected to avoid firing
+  // them at an unexpected time.
+  await BrowserTestUtils.waitForCondition(OtherLazyLoadDataIsReported);
+
+  const snapshot = Services.telemetry.getSnapshotForHistograms("main", false);
+
+  ok(
+    snapshot.parent.FX_LAZYLOAD_IMAGE_PAGE_LOAD_MS.sum > 0,
+    "lazyload image page load telemetry"
+  );
+});